### ISSUE TITLE
awful.widget.layoutbox can never be garbage-collected

### ISSUE THREAD

**Issue by psychon** [awesomeWM/awesome#477]
While debugging the increasing time for redraws in #465, I added a `print(wibox.instances())` and noticed that some wiboxes weren't garbage collected. Adding `collectgarbage("collect")` made this a little better (about half the wiboxes disappeared), but after the "create wibox" test we still had more wiboxes than before.

Turns out this is due to the tooltip that the layoutbox creates. Duh.
I'm not totally sure where this memory leak comes from. For testing I changed the `__mode` of the `local data` in tooltip.lua to "kv", but that did not help.

Some patch for the benchmark:
```diff
diff --git a/tests/test-benchmark.lua b/tests/test-benchmark.lua
index 4b3c319..839188b 100644
--- a/tests/test-benchmark.lua
+++ b/tests/test-benchmark.lua
@@ -64,7 +64,7 @@ local function create_wibox()
 
     -- Create wibox
     local wb = wibox({ width = 1024, height = 20, screen = 1 })
-    wb.visible = true
+    --wb.visible = true
     wb:set_widget(layout)
 
     return wb, textclock
@@ -92,10 +92,16 @@ local function e2e_tag_switch()
     do_pending_repaint()
 end
 
+print(drawin.instances())
 benchmark(create_wibox, "create wibox")
 benchmark(update_textclock, "update textclock")
 benchmark(relayout_textclock, "relayout textclock")
 benchmark(redraw_textclock, "redraw textclock")
 benchmark(e2e_tag_switch, "tag switch")
 
+print(drawin.instances())
+wb = nil
+for i=1, 10 do collectgarbage("collect") end
+print(drawin.instances())
+
 require("_runner").run_steps({ function() return true end })
```
Results without other changes:
```
$ ./tests/run.sh test-benchmark.lua                                                                                        master:1
awesome_log: /tmp/tmp.EnLKq3zc9V/_awesome_test.log
== Running test-benchmark.lua ==
5
        create wibox: 0.100568   sec/iter ( 11 iters, 1.206 sec for benchmark)
    update textclock: 0.00168295 sec/iter (640 iters, 1.993 sec for benchmark)
  relayout textclock: 0.000566826 sec/iter (1815 iters, 1.029 sec for benchmark)
    redraw textclock: 0.0010357  sec/iter (978 iters, 2.789 sec for benchmark)
          tag switch: 0.0390513  sec/iter ( 27 iters, 1.092 sec for benchmark)
17
16
W: awesome: a_glib_poll:291: Last main loop iteration took 8.449660 seconds! Increasing limit for this warning to that value.
```
When I additionally add
```diff
diff --git a/lib/awful/widget/layoutbox.lua b/lib/awful/widget/layoutbox.lua
index 2c0a698..f0cbfa2 100644
--- a/lib/awful/widget/layoutbox.lua
+++ b/lib/awful/widget/layoutbox.lua
@@ -20,7 +20,7 @@ local layoutbox = { mt = {} }
 
 local function update(w, screen, tooltip)
     local layout = layout.getname(layout.get(screen))
-    tooltip:set_text(layout or "[no name]")
+    --tooltip:set_text(layout or "[no name]")
     w:set_image(layout and beautiful["layout_" .. layout])
 end
 
@@ -31,7 +31,7 @@ end
 function layoutbox.new(screen)
     local screen = screen or 1
     local w = imagebox()
-    local tooltip = tooltip({ objects = {w}, delay_show = 1 })
+    --local tooltip = tooltip({ objects = {w}, delay_show = 1 })
 
     update(w, screen, tooltip)
 
```
I get
```
$ ./tests/run.sh test-benchmark.lua                                                                                        master:2
awesome_log: /tmp/tmp.fiF87m4GNC/_awesome_test.log
== Running test-benchmark.lua ==
3
        create wibox: 0.103239   sec/iter ( 11 iters, 1.234 sec for benchmark)
    update textclock: 0.00170447 sec/iter (610 iters, 2.135 sec for benchmark)
  relayout textclock: 0.000639949 sec/iter (1648 iters, 1.055 sec for benchmark)
    redraw textclock: 0.00104936 sec/iter (960 iters, 1.804 sec for benchmark)
          tag switch: 0.0375075  sec/iter ( 27 iters, 1.051 sec for benchmark)
3
2
W: awesome: a_glib_poll:291: Last main loop iteration took 7.629655 seconds! Increasing limit for this warning to that value
```
which shows that some tooltip fails to get garbage collected.

**Comment by psychon** [comment]
It's always these stupid things...

By not adding a tooltip, my "poor leak detection" is just fooled. The problem is that the layoutbox itself cannot be garbage collected. This is because it connects to signals on the tag class and this connection can never be disconnected and keeps everything alive:
```lua
    -- in awful.widget.layoutbox
    tag.attached_connect_signal(screen, "property::selected", update_on_tag_selection)
    tag.attached_connect_signal(screen, "property::layout", update_on_tag_selection)
```

**Comment by psychon** [comment]
All of this should be fixed with #479.
